### ISSUE TITLE
Harvesting from repos (UR Research) and missing 001 and 003

### ISSUE THREAD

**Issue by patrickzurek** [eXtensibleCatalog/xcmetadataservicestoolkit#577]
JIRA issue created by: jbowen
Originally opened: 2011-08-11 04:28 PM

Issue body:
 (nt)

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2011-08-11 04:31 PM

Comment body:

See issue 786.  Below is copied from that issue.


Questions:
1) what should the transformation service do when there are no 001s or 003s?
2) does irplus output holdings?

I'm putting this in 1.0 just so we don't miss it and can prioritize it accordingly.  Is there a better way to do this?  Should we have a queue milestone that we can review on Thursdays?  So that anything in the queue is something that is waiting to be assigned to another milestone?
Assigned to Benjamin Anderson by Benjamin Anderson 8/10/2011 (Yesterday) 9:42 AM
Edited by Randall Cook 8/10/2011 (Yesterday) 2:49 PM Edit
Notified Jennifer Bowen.
Couple thoughts. 

I am not sure about the idea of obtaining MARC records from a non-MARC system, Jennifer should advise on that.  The 001 is the record id for the record in the source system.  What is the relevant ID in IRPlus..is it the handle? a url?  something else?  The 003 for an ILS is typically the LC given Institution Code?  What is this going to be if the repository is not an ILS but rather a digital repository?

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2011-08-17 02:25 PM

Comment body:

I believe that our MARCXML services are going to need to accommodate records harvested without using the OAI Toolkit and that don't have either 001 or 003.  I'm working on a document that tries to pull together how XC uses 001 and 003.  In doing this, I've come to the conclusion that we need to define a 9XX field, and a new XC Schema element, called something like "source repository" to contain the domain name of the original repository.  I'll post that for discussion when I'm done.  The problem is that for digital repositories (such as IR Plus) we can't really shoehorn this data into the 003, and such systems don't define an 001 anyway.  Bridget's question about increasing the number of characters for the 003 on Google Code brought up this issue as well. 

To answer Ben's questions below:
1.  The Transformation Service will need to process records that don't have 001 and 003.  This may require some other changes to both Transformation and Normalization, which I'll start to spec out.
2.  No - IRplus does not output holdings.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2011-08-18 08:40 PM

Comment body:

This has brought up several issues.  I went through and documented how various parts of the software are (or should be) using both 001 and 003 fields in this document:  http://docushare.lib.rochester.edu/docushare/dsweb/Get/Document-48280 (in the Docushare Metadata Services folder).  I'd like feedback on this.  Basically I think that unless a bib record has separate holdings that we should not require 001 and 003, since many repositories will not supply them.  The exception is the OAI Toolkit, which needs both.

If not all records will have an 003, then we have another need to be able to track the source repository for a record so that we can generate a facet value for the repository.  I am proposing that we add a new XC Schema element (and a placeholder 9XX field), which will require changes to the other services to map the data, and I'll add that as a separate FogBugz case.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2011-08-24 08:43 PM

Comment body:

I've now made changes to the Normalization Services documentation and config file http://docushare.lib.rochester.edu/docushare/dsweb/Get/Document-48283/ to address this (the doc also addresses other stuff as well, such as Fogbugz 766).  Dave should check the various scenarios in this document to see if they agree with what we discussed.

**Comment by patrickzurek** [comment]
JIRA Coment by user: dlindahl
JIRA Timestamp: 2011-09-07 05:31 PM

Comment body:

I added some comments to the document.  I think it is good.  Just a couple questions - probably clarifications.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2011-09-09 09:03 PM

Comment body:

After looking at your requests for clarifications I decided to rewrite this to make it much clearer.  In order to make sure I was documenting things correctly, I also turned the "which services need 001 and 003"? document into a spreadsheet - it is posted in the Normalization Changes Needed for Aggregation folder in Docushare with the new version of this document.
http://docushare.lib.rochester.edu/docushare/dsweb/View/Collection-6820  

Dave, please take another look.  Randy, would probably help for you to look through it too to see if you see anything that looks incorrect.

Oh, and this document brought up the need for error cases for Aggregation merging, so I added some to that document too.  Not sure what Fogbugz error that one is offhand.

**Comment by patrickzurek** [comment]
JIRA Coment by user: dlindahl
JIRA Timestamp: 2011-09-14 07:52 PM

Comment body:

I am going to close this, since it now refers to the same set of documents that issue 766 refers too, and it is confusing having two fogbugz cases about a set of four documents.  Just dealing with the 4 documents is enough to think about.  Done.  766 remains open.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2011-11-14 06:37 PM

Comment body:

Reopening this and un-bundling some functionality from the other issues. 

Since we have members of the community wanting to be able to harvest MARCXML data from other repositories and run it through MARC Transformation NOW, we feel we should fix the issue with Transformation ASAP and schedule the embedded holdings cleanup (which is complicated) later. 

For now, make sure that Transformation doesnâ€™t require the 001 now, which is a show-stopper for people.

I believe that Transformation currently uses the 003 to determine which steps to perform and there is a longer term solution proposed in other isses that involves use of a 9xx field, but that will be more difficult to fix.  However, there is a work-around for this because libraries can turn on the step in Normalization to supply an 003 if their records donâ€™t contain 003s.  That might be a problem if their institution doesnâ€™t have a MARC code, but itâ€™s easy enough to get one or make one up, and they should at least be able to use the MST in the meantime with this change.  


To test this, you should test with both UR Research (check with Nate) and at least one of the repos that the Spanish folks are trying to use, such as below.

Status: Active
Repository: Univ autn Madrid
Repository URL: http://digitool-uam.greendata.es/OAI-PUB
Schema: marc21
Sets: Univ-autn-Madrid ,  Univ-autn-Madrid:driver
Created at: 2011-10-10 22:35:24.0
Updated at: 2011-11-10 12:46:08.0
OAI datestamp:
OAI identifier: oai:digitool-uam.greendata.es:11532

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2011-11-15 02:32 PM

Comment body:

To start with, I tried to harvest from URResearch and process the results through our services.  There were a number of exceptions preventing the processing through transformation.  These were all around checking whether bibs were received or any holding was waiting on any bibs.  Made this code work when there is no 003.  For this repository, there seemed to be no fallout from the lack of 001's.  After the change for transformation for 003's, all records ran through our services (in the small 500 record sample-set I ran).

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2011-11-15 06:38 PM

Comment body:

Tested on 


I investigated harvesting from the following digital repository:
Repository: Univ autn Madrid
Repository URL: http://digitool-uam.greendata.es/OAI-PUB
Schema: marc21
Sets: Univ-autn-Madrid ,  Univ-autn-Madrid:driver
Created at: 2011-10-10 22:35:24.0
Updated at: 2011-11-10 12:46:08.0
OAI datestamp:
OAI identifier: oai:digitool-uam.greendata.es:11532

I get the following error on a number of records:

15 Nov 2011 10:05:49,527 xc.mst.services.transformation.TransformationService:511 ERROR [SERVICE_1035] - error processing record with id:3506
java.lang.NumberFormatException: For input string: "FMT"
    at java.lang.NumberFormatException.forInputString(NumberFormatException.java:48)
    at java.lang.Integer.parseInt(Integer.java:449)
    at java.lang.Integer.parseInt(Integer.java:499)
    at xc.mst.bo.record.SaxMarcXmlRecord.startElement(SaxMarcXmlRecord.java:165)
    at org.apache.xerces.parsers.AbstractSAXParser.startElement(Unknown Source)

Note that our xml parser that is complaining about this tag in the record, should we support this sort of non-numeric tag?

<record xmlns="http://www.loc.gov/MARC21/slim" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.loc.gov/MARC21/slim http://www.loc.gov/standards/marcxml/schema/MARC21slim.xsd">


  <leader>01145cam  2200253 i 4500</leader>
  <controlfield tag="FMT">BK</controlfield>






Jennifer's reply:

A non-numeric tag is totally non-standard. I don't think we should support it.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2011-11-22 08:54 PM

Comment body:

created gc issue 324

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2011-11-23 07:48 PM

Comment body:

I reopened this because I created a new version of the Docushare doc on how XC uses 001 and 003.  http://docushare.lib.rochester.edu/docushare/dsweb/View/Collection-4969

I will move this to Google docs and link to it on the wiki when it looks accurate, and refer Salva to it.  Randy and John, please have a look. I tried to include what's happening now AND what we plan to do in the future.

(Randy, have a look at the Drupal part - is what I wrote about NCIP accurate?)

If this needs to be a new Fogbugz issue please move it...

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2011-11-30 03:12 PM

Comment body:

Another new version of the 001/003 document now uploaded to Docushare.  I found lots of problems so rewrote the Norm service section completely and got rid of the Aggregation section since it wasn't quite right. Randy and John, please review and if it looks ok now I'll move it to the wiki.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbrand
JIRA Timestamp: 2012-02-15 02:08 PM

Comment body:

This document looks good to me.  Jennifer asked for comments from Randy and myself, so now assigning to Randy.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-02-15 03:40 PM

Comment body:

Actually, sending to Jennifer.  She is doing another "rethink" on this, and then she, Dave and I will meet to discuss, possibly tomorrow.

Several issues are all interrelated around 001 and 003 work, including this, embedded holdings, and the changes needed to get the services to support MARC Aggregation.

**Comment by patrickzurek** [comment]
JIRA Coment by user: rcook
JIRA Timestamp: 2012-03-15 02:28 PM

Comment body:

John question on your 11/15 comment.  Does this mean, that as of right now:

1.  The MARC Norm service can process records successfully with neither 001 nor 003?  If this does not currently work when neither are present, what would it take to make that change (e.g. make it on/off)
2.  The MARC Transform service can process records sucessfully with neither 001 nor 003?  

John/Chris - can one of you very clearly articulate for each MARC service, what happens when:

1. There is an 001 and there is no 003
2. There is neither an 001 nor an 003?

I think we are all good with what happens with there is an 001 and an 003 or when we use the supply an 003 step.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2012-03-16 02:02 PM

Comment body:

This sounds good to me. Thanks for confirming, John.  I'll let you know if I have further questions.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2012-03-16 02:10 PM

Comment body:

For Transformation:  good to know that the service will RUN without 003, but I believe that it will not correctly map the 9XX fields created by Normalization, which look for the presence of the org code in $5. So it really needs it to run properly given the current configuration. I'll try to explain that in the documentation that I'm writing.  I think the service is doing what it is supposed to do.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2012-08-24 11:43 AM

Comment body:

Now that MST-469 has been implemented (changes to 001 and 003 fields) we should do a test harvest from IR+ and process the records through a separate Normalization Service that is configured to supply both 001 and 003 fields for these records.  We might want to consult with [~dlindahl]when setting this up to make sure we have set up the 001 and 003 options as intended.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2012-08-24 01:34 PM

Comment body:

I created a new issue RCL-27 to deal with a new Norm service for IR+ since it is really an RCL issue rather than an issue for the MST.

**Comment by patrickzurek** [comment]
JIRA Coment by user: jbowen
JIRA Timestamp: 2012-08-24 01:37 PM

Comment body:

The 001/003 issues were fixed with MST-469; the IR+ stuff is now all under RCL issues.

**Comment by patrickzurek** [comment]
Issue resolved: 2012-08-24 01:37 PM
